Add module GHA doc (#528)
AlexsLemonade/OpenScPCA-analysis: docs/ensuring-repro/workflows/run-module-gha.md

# Automated module testing

> To maintain module functionality over time, we use [GitHub Actions](https://docs.github.com/en/actions) (GHAs) to periodically run each module ("module testing GHA") with the goal of testing that it runs to completion without errors.

Reviewer: ```suggestion
To maintain module functionality over time, we use [GitHub Actions](https://docs.github.com/en/actions) (GHAs) to periodically run each module ("module testing GHA") with the goal of testing that the module code runs to completion without errors.
```

!!! info
    For more information about how we run modules to generate official results and OpenScPCA releases, please see our documentation on the `OpenScPCA-nf` workflow. <!-- openscpca-nf STUB_LINK -->

> Module testing GHAs are automatically run under these two circumstances:
> 
> - When a pull request is filed with changes to any module files
>     - This GHA will need to pass without errors for [pull requests](../../contributing-to-analyses/pr-review-and-merge/index.md) to be approved
> - All modules are periodically run on a schedule to ensure they continue to pass tests as OpenScPCA grows

Reviewer: My original motivation to changes here was to make the two bullets parallel, but I'm not totally happy with the result... maybe you can see where I am going?
```suggestion
Module testing GHAs are automatically run in two circumstances:

- When a pull request is filed with changes to any module files
    - This GHA will need to pass without errors for [pull requests](../../contributing-to-analyses/pr-review-and-merge/index.md) to be approved
- On a periodic schedule
  - This ensures that changes in data or other code do not break tests within each module
```

Author: I thought your proposed structure was good here!

> To make GHAs run efficiently, we strongly encourage and generally require, when applicable, that GHAs run the module code using the [simulated test data](../../getting-started/accessing-resources/getting-access-to-data.md#accessing-test-data).

Reviewer: Simplifying.

```suggestion
To make GHAs run efficiently, the tests should run the module code with the [simulated test data](../../getting-started/accessing-resources/getting-access-to-data.md#accessing-test-data).
```

> This means that it's important to write your module code with sufficient flexibility to allow for test data to be used.
> For example, when reading input data into scripts, you should use arguments for input data files, allowing for test data paths to be used in these module testing GHAs.

Reviewer: The test paths are designed to match real data paths, so no changes should be needed in most cases. 

```suggestion
This means that it's important to write your module code with sufficient flexibility to allow for test data to be used.
You should read in files from the `data/current` directory, which will be automatically directed to test data during GHA test runs. 
```

If you want to add more detail here, it might be that things like checks for numbers of cells might need to be adjusted, but these kinds of things will be specific to individual modules and I do not think it is something we will be able to have any general instructions about. 

I think a more general thing to say (maybe in the next section?) is that it is helpful for testing to have a single entry point to run the whole module. Or at least a set of well-labelled scripts. You do mention this later, but it probably makes sense to talk about here too.

Author: Added a couple spots about data to clarify this

## Writing a module testing GHA

!!! tip
    The Data Lab will generally maintain and write module testing GHAs, but you are welcome to do so as well if you are interested!
    See this GitHub documentation to learn about [workflow syntax for GHAs](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions).

When you [create a new module](../../contributing-to-analyses/analysis-modules/creating-a-module.md), a GHA workflow file is created in the file `.github/workflows/run_{module-name}.yml`.
This initial file is inactive, meaning it will not run automatically run on the two aforementioned triggers.
As analysis module begins to mature over time, the Data Lab staff will activate this workflow file so the module can be regularly tested.

### GHA steps

Each module testing GHA is initially created with these steps, which should be updated to reflect the given module's needs:

- Checkout the repository
- Download test data

> - Setup the module environment

Reviewer: ```suggestion
- Set up the module environment
```

> - Setup the module environment

Reviewer: We may want to reverse the order of these two, when I think about it. Mostly because environment setup tends to be the harder part so we would rather fail with that first.

Author: I filed https://github.com/AlexsLemonade/OpenScPCA-analysis/issues/529 for later

    - Depending on [the flags used when creating your module](../../contributing-to-analyses/analysis-modules/creating-a-module.md#module-creation-script-flags), this will steps steps needed to install the [`renv` and/or conda environment](../managing-software/index.md) from existing environment files (`renv.lock` and/or `conda-lock.yml`, respectively)
- Run the analysis module
    - Generally, this will involve calling the [module's run script](../../contributing-to-analyses/analysis-modules/running-a-module.md)

> As an analysis module matures, the GHA should instead be updated to run the analysis in the module's Docker image, rather than using the `renv` and/or conda environment files.

Reviewer: ```suggestion
As an analysis module matures, the GHA will be updated to run the analysis in the module's Docker image, rather than using the `renv` and/or conda environment files.
```

- Checkout the repository
- Download test data

>     - After this step the `data/current` directory will point to the test data, which ensures the module GHA to runs using the test data.

Reviewer: We probably want to say something about downloading the correct values here?
I used stub links just because I can't find the correct links easily in this interface...
```suggestion
    - Use the [`download-data.py`](#STUB_LINK) and/or [`download-results.py`](#STUB_LINK) scripts to specify the set of input files you need, with the `--test-data` flag to specify downloading the test data.
    - After this step the `data/current` directory will point to the test data, which ensures the module GHA to runs using the test data.
```
